arch/arm/core/aarch64/userspace.S

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2020 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <offsets_short.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+#include "macro_priv.inc"
+
+_ASM_FILE_PROLOGUE
+
+/*
+ * size_t arch_user_string_nlen(const char *s, size_t maxsize, int *err_arg)
+ */
+
+GTEXT(z_arm64_user_string_nlen_fault_start)
+GTEXT(z_arm64_user_string_nlen_fault_end)
+GTEXT(z_arm64_user_string_nlen_fixup)
+
+GTEXT(arch_user_string_nlen)
+SECTION_FUNC(TEXT, arch_user_string_nlen)
+
+	mov	x3, x0
+	mov	x0, #0
+	mov	x4, #0
+
+strlen_loop:
+
+	cmp	x0, x1
+	beq	strlen_done
+
+z_arm64_user_string_nlen_fault_start:
+	ldrb	w5, [x3, x0]
+z_arm64_user_string_nlen_fault_end:
+	cbz	x5, strlen_done
+
+	add	x0, x0, #1
+	b	strlen_loop
+
+z_arm64_user_string_nlen_fixup:
+	mov	x4, #-1
+	mov	x0, #0
+
+strlen_done:
+	str	w4, [x2]
+	ret
+
+/*
+ * int arch_buffer_validate(void *addr, size_t size, int write)
+ */
+
+GTEXT(arch_buffer_validate_fault_start)
+GTEXT(arch_buffer_validate_fault_end)
+GTEXT(arch_buffer_validate_fixup)
+
+GTEXT(arch_buffer_validate)
+SECTION_FUNC(TEXT, arch_buffer_validate)
+
+	add	x1, x1, x0
+
+arch_buffer_validate_fault_start:
+	ldtrb	w3, [x0]
+	cbz	w2, arch_buffer_validate_fault_end
+	sttrb	w3, [x0]
```

Comment:
I see what this is intended to do; I am rather hesitant about "null" writebacks, as there's a lot that can go wrong.

Perhaps instead use the `AT` instructions? They will trigger a synchronous abort if it's not allowed.